# Image Recognition (Cat & Dog dataset)
<b> In our dataset, we have all the images of cats and dogs in training as well as in the test set folders. We are going to train our CNN model on 4048 images of cats as well as 4000 images of dogs, each respectively that are present in the training set followed by evaluating our model with the new 1000 images of cats and 1000 images of dogs, each respectively in the test set on which our model was not trained. So, we are actually going to build and train a Convolutional Neural network to recognize if there is a dog or cat in the image.

### importing all the necessary libraries.
    
Here we are importing the TensorFlow, keras library and actually the preprocessing module by Keras library. And then, we will import the image sub-module of the preprocessing module of the Keras library, which will allow us to do image pre-processing.

In [1]:
import tensorflow as tf
import keras
from keras.preprocessing.image import ImageDataGenerator

### Data Pre-processing
<b> Preprocessing the Training set
    
    

In [2]:
train_datagen = ImageDataGenerator(rescale=1./255, shear_range=0.2, zoom_range=0.2)

In the above, We have created an object of 'train_datagen' of the ImageDataGenerator class that represents the tool that will apply all the transformations on the images of the training set, such that the **rescale** argument will apply feature scaling to each and every single one the pixel by dividing their value 255 as each pixel take a value between 0 and 255, which is really necessary for neural networks and the rest are the transformations that will perform image augmentation on the training set images so as to prevent the overfitting. **shear_range:** Shear Intensity (Shear angle in counter-clockwise direction in degrees). **zoom_range:** Range for random zoom. [lower, upper] = [1-zoom_range, 1+zoom_range]`.


After this, we are going to connect the '**train_datagen**' object to the training set, and to do this, we are going to import the training set, which can be done as given below. Here training_set is the name of the training set that we are importing in the notebook, and then we indeed take our train_datagen object so as to call the method of **ImageDataGenerator class**. The method that we will call is the **flow_from_directory** that will help to connect the image augmentation tool to the image of the training set. we will pass the following parameter;


- The first parameter is the **path** leading to the training set.


- The next parameter is the **target_size**, which is the final size of the images when they will be fed into the convolutional neural network.


- The third one is the **batch_size**, which relates to the size of the batches, i.e., the total number of images we want to have in each batch. We have chosen 32, which is the classic default value.


- Lastly, we will classify the **class mode** to be either binary or categorical. Since we are looking for a binary outcome, so will choose binary class mode.



In [3]:
training_set = train_datagen.flow_from_directory('dataset/training_set', target_size = (64, 64),
                                                         batch_size = 32, class_mode = 'binary')

Found 8048 images belonging to 2 classes.


From the above, we have got the output from the above image that indeed we imported and preprocessed with the data augmentation; 8048 images belonging to 2 classes.

Now we are checking the those 2 classes by using 'class_indices' attribute, i.e., dogs and cats. The 'class_indices' will return the dictionary that contains the mapping from class names to class indices.

In [4]:
training_set.class_indices

{'cats': 0, 'dogs': 1}

<b> Preprocessing the Test set
    
After we are done with preprocessing the training set, we will further move on to preprocessing the test set. We will again take the ImageDataGenerator object to apply transformations to the test images, but here we will not apply the same transformations as we did in the previous step. However, we need to rescale their pixels the same as before because the future predict method of CNN will have to be applied to the same scaling as the one that was applied to the training set.    

In [5]:
test_datagen = ImageDataGenerator(rescale = 1./255)

After this, we are going to connect the '**test_datagen**' object to the **testing set**, and to do this, we are going to import the **testing set**, which can be done as given below. Here **test_set** is the name of the test set that we are importing in the notebook, and then we indeed take our **test_datagen**, which will only apply if it is going to the pixels of the test set images. Then we call the same '**flow_from_directory**' function to access the test set from the directory. Then we will need to have the same **target_size, batch_size, and class_mode** as used in the previous step.

In [6]:
testing_set = test_datagen.flow_from_directory('dataset/test_set', target_size = (64, 64),  
                                                batch_size = 32, class_mode = 'binary')  

Found 2000 images belonging to 2 classes.


We can see from the above image, which we got after running Preprocessing the Test Set cell, that 2000 images belong to 2 classes. Instead of applying image augmentation, we have only applied feature scaling.

Now we are checking the those 2 classes by using 'class_indices' attribute, i.e., dogs and cats. The 'class_indices' will return the dictionary that contains the mapping from class names to class indices.

In [7]:
testing_set.class_indices

{'cats': 0, 'dogs': 1}

### Modelling / Building the Convolution Neural Network (CNN)
Hare we are going to build the convolutional neural network and, more specifically, the whole architecture of the artificial neural network. So, it is actually going to start the same as with our artificial neural network because the convolutional neural network is still a sequence of layers.


Therefore, we are going to initialize our CNN with the same class, which is the sequential class.

<b> Initializing the CNN
    

In [8]:
# importing Sequential clas from keras.models library.
from keras.models import Sequential 

# initialize the CNN - Sequential class as 'classifier'
classifier = Sequential()

After this, we will step by step use the add method to add different layers, whether they are convolutional layers or fully connected layers, and in the end, the output layer. So, we are now going to successfully use the add method starting with the step1: convolution.

<b> Step1: Convolution
    
Here we will first take the 'classifier' object or the convolutional neural network from which we will call the add method to add our very first convolutional layer, which will further be an object of a certain class, i.e., **Conv2D** class. And this class, just like the dense class that allows us to build a fully connected layer, belongs to the same module, which is the layer module from the Keras library.

    
Inside the class, we are going to pass four important parameters, which are as follows:

- The first parameter is the **filters**, which is the number of feature detectors that we want to apply to images for feature detection.
    
    
- The **kernel_size** is exactly the size of the feature detector, i.e., the number of rows, which is also the number of columns.
    
    
- The third one is the **activation** but here we are not going to keep the default value for the activation parameter corresponding to the activation function, because indeed as long as we don't reach the output layer, we rather want to get a rectifier activation function. That is why we will choose the ReLU parameter name once again as it corresponds to the rectifier activation function.
    
    
- Lastly, the **input_shape** parameter because it is necessary to specify the input shape of inputs. Since we are working with the colored images, so the input_shape will be [64, 64, 3].
    
    

In [9]:
# importing Conv2D clas from keras.layers library.
from keras.layers import Conv2D

# Convolution operation and relu
classifier.add(Conv2D(input_shape=[64,64,3], filters=32, kernel_size=3, activation='relu'))

<b> Step2: Pooling

Next, we will move on to applying pooling, and more specifically, if we talk about, we are going to apply the max pooling, and for that, we will again take cnn object from which we are going to call our new method. Since we are adding the pooling layer to our convolutional layer, so we will again call the add method, and inside it, we will create an object of a max-pooling layer or an instance of a certain class, which is called **MaxPooling2D** class. Inside the class, we will pass **pool_size** and **strides** parameters. 


- pool_size: integer or tuple of 2 integers, window size over which to take the maximum. (2, 2) will take the max value over a 2x2 pooling window. If only one integer is specified, the same window length will be used for both dimensions.


- strides: Integer, tuple of 2 integers, or None. Strides values. Specifies how far the pooling window moves for each pooling step. If None, it will default to pool_size.

In [10]:
# importing MaxPooling2D clas from keras.layers library.
from keras.layers import MaxPooling2D

# applying the pooling classifier
classifier.add(MaxPooling2D(pool_size=2, strides=2))

<b> Step3: Flattening
    
In the third step, we will undergo flattening the result of these convolutions and pooling into a one-dimensional vector, which will become the input of a fully connected layer neural network. We will start with again taking our '**classifier**' object from which we will call the **add** method because the way we are going to create that flattening layer is once again by creating an instance of the **Flatten** class, such that Keras will automatically understand that this is the result of all these convolutions and pooling, which will be flattened into the one-dimensional vector.

    
So, we just need to specify that we want to apply flattening and to do this we will have to call once again the layers module by the Keras library from which we are actually going to call the flatten class, and we don't need to pass any kind of parameter inside it.    

In [12]:
# importing Flatten clas from keras.layers library.
from keras.layers import Flatten

# flattening the result
classifier.add(Flatten())

<b> Step4: Full Connection
    
In step 4, we are exactly in the same situation as before building a fully connected neural network. So, we will be adding a new fully-connected layer to that flatten layer, which is nothing but a one-dimensional vector that will become the input of a fully connected neural network. And for this, we will again start by taking a '**classifier**' neural network from which we are going to call the **add** method because now we are about to add a new layer, which is a fully connected layer that belongs to **keras.layers.** But this time, we will take a **Dense** class followed by passing **units**, which is the number of hidden neurons we want to have into this fully connected layer and **activation function** parameter.   
    
    

In [13]:
# importing Dense clas from keras.layers library.
from keras.layers import Dense

# hidden layer with 128 neurons
classifier.add(Dense(units=128, activation='relu'))

<b> Step5: Output Layer
    
Here we need to add the final output layer, which will be fully connected to the previous hidden layer. Therefore, we will call the **Dense** class once again in the same way as we did in the previous step but will change the value of the input parameters because the numbers of units in the output layer are definitely not 128. Since we are doing binary classification, it will actually be one neuron to encode that **binary class** into a 'cat' or 'dog'. And for the activation layer, it is recommended to have a **sigmoid** activation function. Since this is a binary classification problem, so we are using **sigmoid** activation function    

In [14]:
# output layer with 1 neuron
classifier.add(Dense(units=1, activation='sigmoid'))

### Training the CNN
<b> Training the CNN model with train data & Testing the model with test data 
    
Now we are going to train our CNN over 25 **epochs**, and at each epoch, we will actually see how our model is performing on our test set images. This is a different kind of training as we did before because we always used to separate the training and evaluation, but here this will happen at the same time as we are making some specific application, i.e., computer vision.  
    
<b> Compiling the CNN
    
Now we are going to compile the CNN, which means that we are going to connect it to an optimizer, a loss function, and some metrics. As we are doing once again a binary classification, so we are going to compile our CNN exactly the same way as we complied our ANN model because indeed, we are going to choose once again **adam** optimizer to perform **stochastic gradient descent** to update the weights in order to reduce the loss error between the predictions and target. Then we will choose the same loss, i.e., the **binary_crossentrophy** loss because we are doing exactly the same task binary classification. And then same for the metrics, we will choose **accuracy** metrics because it is the most relevant way to measure the performance of the classification model, which is exactly our case of CNN.

So, we will take our 'classifier' object from which we will be calling the compile method that will take as input the optimizer, loss function, and the metrics.    

In [15]:
# Compiling the CNN
classifier.compile(optimizer='adam', loss='binary_crossentropy', metrics='accuracy')

<b> Training the CNN on the Training set and evaluation on the Test set
    
After the compilation, we will train the CNN on the training set followed by evaluating at the same time on the test set, which will not be exactly the same as before but will be somewhat similar. Basically, the first two steps are always the same, i.e., in the first step, we will take 'classifier' followed by taking the fit method in the second step that will train the 'classifier' on the training set. Inside it, we will pass the following parameters:

- The first parameter is the set, which is off course the dataset (**training set**) on which we are going to train our model, and the name for that parameter is simply x.
    
    
- The second parameter is the difference with what we did before. So, it has to do, of course with the fact that we are not only training the CNN on the training set but also evaluating it at the same time on the test set. And that is what exactly our second parameter corresponds to, so we will be specifying here the **validation data (testing set)**, which is the set on which we want to evaluate our CNN.
    
    
Lastly, the **epochs** parameter, which is the number of epochs. Here we are choosing 25 epochs to converge the accuracy not only on the training set but also on the test set.

In [16]:
classifier.fit(x = training_set, validation_data = testing_set, epochs = 25)

Epoch 1/25
252/252 [==============================] - 153s 600ms/step - loss: 0.6805 - accuracy: 0.6221 - val_loss: 0.5811 - val_accuracy: 0.6935
Epoch 2/25
252/252 [==============================] - 113s 448ms/step - loss: 0.5815 - accuracy: 0.6873 - val_loss: 0.5581 - val_accuracy: 0.7115
Epoch 3/25
252/252 [==============================] - 105s 418ms/step - loss: 0.5458 - accuracy: 0.7184 - val_loss: 0.5362 - val_accuracy: 0.7205
Epoch 4/25
252/252 [==============================] - 129s 512ms/step - loss: 0.5282 - accuracy: 0.7299 - val_loss: 0.5533 - val_accuracy: 0.7310
Epoch 5/25
252/252 [==============================] - 95s 374ms/step - loss: 0.4935 - accuracy: 0.7565 - val_loss: 0.5395 - val_accuracy: 0.7305
Epoch 6/25
252/252 [==============================] - 96s 379ms/step - loss: 0.4780 - accuracy: 0.7681 - val_loss: 0.7074 - val_accuracy: 0.6575
Epoch 7/25
252/252 [==============================] - 88s 348ms/step - loss: 0.4601 - accuracy: 0.7824 - val_loss: 0.5420 - va

<b> From the image given above, we would have ended up with an accuracy of around 95% on the training set, which clearly indicates overfitting and lower accuracy here on the test set around 74%. 

<b> So now we are going to rebuild the model with 'Dropout' function with '0.1'.

### Rebuilding the model

In [17]:
# importing Sequential clas from keras.models library.
from keras.models import Sequential 

# initialize the CNN - Sequential class as 'classifier'
model = Sequential()

# importing Conv2D clas from keras.layers library.
from keras.layers import Conv2D, Dropout

# Convolution operation and relu
model.add(Conv2D(input_shape=[64,64,3], filters=32, kernel_size=3, activation='relu'))

# importing MaxPooling2D clas from keras.layers library.
from keras.layers import MaxPooling2D

# applying the pooling classifier
model.add(MaxPooling2D(pool_size=2, strides=2))
model.add(Dropout(0.1))                              # Dropout with 10% 

# importing Flatten clas from keras.layers library.                     
from keras.layers import Flatten

# flattening the result
model.add(Flatten())
model.add(Dropout(0.1))                              # Dropout with 10% 

# importing Dense clas from keras.layers library.
from keras.layers import Dense

# hidden layer with 128 neurons
model.add(Dense(units=128, activation='relu'))
model.add(Dropout(0.1))                              # Dropout with 10% 

# output layer with 1 neuron
model.add(Dense(units=1, activation='sigmoid'))
model.add(Dropout(0.1))                              # Dropout with 10% 

# Compiling the CNN
model.compile(optimizer='adam', loss='binary_crossentropy', metrics='accuracy')

# Training the CNN on the Training set and evaluation on the Test set
model.fit(x = training_set, validation_data = testing_set, epochs = 25)

Epoch 1/25
252/252 [==============================] - 144s 559ms/step - loss: 1.3387 - accuracy: 0.5859 - val_loss: 0.6409 - val_accuracy: 0.6195
Epoch 2/25
252/252 [==============================] - 104s 412ms/step - loss: 1.2841 - accuracy: 0.6553 - val_loss: 0.5663 - val_accuracy: 0.7135
Epoch 3/25
252/252 [==============================] - 101s 401ms/step - loss: 1.2929 - accuracy: 0.6644 - val_loss: 0.5979 - val_accuracy: 0.7040
Epoch 4/25
252/252 [==============================] - 104s 415ms/step - loss: 1.2669 - accuracy: 0.6827 - val_loss: 0.5957 - val_accuracy: 0.6730
Epoch 5/25
252/252 [==============================] - 87s 347ms/step - loss: 1.2916 - accuracy: 0.6940 - val_loss: 0.5617 - val_accuracy: 0.7090
Epoch 6/25
252/252 [==============================] - 88s 349ms/step - loss: 1.1746 - accuracy: 0.7107 - val_loss: 0.5335 - val_accuracy: 0.7440
Epoch 7/25
252/252 [==============================] - 89s 354ms/step - loss: 1.2206 - accuracy: 0.7249 - val_loss: 0.5414 - va

<b> From the image given above, it can be seen that we ended with 79% of final accuracy on the training set and final accuracy of 74% on the test set. Based on this, we conclude that we have got good model.

### Evaluation:
<b> Making a single prediction
    
Here we will make a single prediction, based on the rebuild model. Now we are performing evaluation on the two separate images of this single prediction folder for which our model will have to recognize for both the dog and cat, respectively.

    
To display the image, '**PIL**' library is using an '**Image**' class within it. The image module inside pillow package contains some important inbuilt functions like, load images or create new images, etc.
    
   

In [19]:
import numpy as np
from PIL import Image

To load the image, we simply import the '**Image**' module from the 'PIL' and call the **Image.open()**, passing the image filename. Which returns a value of the Image object data type. Any modification we make to the image object can be saved to an image file with the save() method. The image object we received using Image.open(), later can be used to **resize** image manipulation method calls on this Image object.


But to make our first **test _set** image accepted by the predict method, we need to convert the format of an image into an array because the predict method expects its input to be a 2D array. And we will do this with the help of another function of the image preprocessing module, i.e., **np.array()** function, which indeed converts PIL image instance into a NumPy array that is exactly the format of array expected by the predict method. We are going to use the np.array(), and inside, it will take the test_size image in PIL format that we are looking forward to convert it into the NumPy array format.


Next, we will add an extra dimension, which will correspond to the batch that will contain that image into the batch, and it can be simply done by updating our test image by adding extra dimensions corresponding to batch. And the way to do it is with NumPy as the NumPy arrays can be easily manipulated, so we will first call the NumPy from which we will call this function that allows exactly to add a fake dimension, or we can say a dimension corresponding to the batch, which is called **expand_dims** function inside of which we will input the image to which we want to add this extra dimension corresponding to the batch followed by adding an extra argument, which is an axis that we have to set equal to **zero**. That is why the dimension of a batch that we want to add to our image will be the first dimension.


After this, we are going to call the predict method on that '**test_image**' and saving the same result as 'result' object.

After this, inside the batch, we are going to get access to the first element of the batch that corresponds to the prediction of that same cat_or_dog_1 image. As we are dealing with a single image, so a single prediction is needed, and to get that, we will need to get inside the batch of index zero, the first and only prediction once again, which has a [0] index. So, that is how we get our prediction by first accessing the batch followed by accessing the single element of the batch, and if that prediction equals to one, then we already know that it corresponds to the dog, then we will create a new variable which we will call as prediction and will set that prediction variable equals to the dog. Likewise, in the else condition, if the result prediction equals to 1, then the prediction will be a cat. Now we will wrap it up by simply printing the prediction.

In [22]:
# Load the data
test_image = Image.open('dataset\single_prediction\cat_or_dog_1.jpg')

# Data Preprocessing
test_image = test_image.resize((64, 64))
test_image = np.array(test_image)
test_image = np.expand_dims(test_image, axis=0)

# Prediction
result = model.predict(test_image)

# Evaluation
if result[0][0]==1:         # If result=1 then print Dog else Cat
    print('Dog')
else:
    print('Cat')

1/1 [==============================] - 1s 675ms/step
Dog


<b> We can see our Convolution Neural Network predicted that there is a dog inside the image. So, it can be concluded that our first test is passed successfully.
    
    
<b> Now we will check for the other image which is of the cat, so for that we will need to deploy our model on this single image and check that indeed, our CNN returns a cat. To do this, we need to change the name here, i.e. cat_or_dog_2.jpg and then play this cell again by clicking on the Run button.

In [23]:
# Load the data
test_image = Image.open('dataset\single_prediction\cat_or_dog_2.jpg')

# Data Preprocessing
test_image = test_image.resize((64, 64))
test_image = np.array(test_image)
test_image = np.expand_dims(test_image, axis=0)

# Prediction
result = model.predict(test_image)

# Evaluation
if result[0][0]==1:         # If result=1 then print Dog else Cat
    print('Dog')
else:
    print('Cat')

1/1 [==============================] - 0s 60ms/step
Cat


<b> So, it's clear now that our CNN model is successful in predicting cat in the output of the console. Hence our CNN got all the answers correct.